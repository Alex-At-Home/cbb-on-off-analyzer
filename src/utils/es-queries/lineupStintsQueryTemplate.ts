import { commonRuntimeMappings } from "./commonRuntimeMappings";
import { commonTeamQuery, buildQueryFiltersBoolArray } from "./commonTeamQuery";
import { QueryUtils } from "../QueryUtils";
import { CommonFilterParams } from "../FilterModels";

export const lineupStintsQuery = function (
  params: CommonFilterParams,
  lastDate: number,
  publicEfficiency: any,
  lookup: any,
  avgEfficiency: number,
  hca: number
) {
  return {
    ...commonRuntimeMappings(params, lastDate, publicEfficiency, lookup),
    _source: {
      includes: [],
      excludes: ["agent", "host", "log", "ecs"],
    },
    sort: ["start_min"],
    size: 200,
    query: {
      bool: {
        filter: [],
        minimum_should_match: QueryUtils.invertedQueryMode(params) ? 1 : 0,
        //(special internal invert mode for getting linueps for on/off)
        should: QueryUtils.invertedQueryMode(params)
          ? buildQueryFiltersBoolArray(
              params.invertBaseQueryFilters,
              params.gender,
              params.year,
              lastDate
            )
              .map((clause) => {
                return {
                  bool: {
                    must_not: [clause],
                  },
                };
              })
              .concat([
                {
                  query_string: {
                    query: `NOT (${QueryUtils.basicOrAdvancedQuery(
                      params.invertBase,
                      "NOT *"
                    )})`,
                  },
                },
              ] as any[])
          : [],
        must_not: [],
        must: [
          commonTeamQuery(params, lastDate, publicEfficiency, lookup),
          {
            query_string: {
              query: `${QueryUtils.basicOrAdvancedQuery(
                params.baseQuery,
                "*"
              )}`,
            },
          },
        ],
      },
    },
  };
};
