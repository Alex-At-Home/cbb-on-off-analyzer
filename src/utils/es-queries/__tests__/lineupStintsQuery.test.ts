import { lineupStintsQuery } from "../lineupStintsQueryTemplate";

describe("lineupStintsQuery", () => {
  test("lineupStintsQuery", () => {
    const test = lineupStintsQuery(
      {
        team: "TestTeam",
        year: "2019",
        minRank: "10",
        maxRank: "100",
        baseQuery: "base",
      },
      0,
      { team: { stats: 0 } },
      {
        name1: "name1b",
      },
      100.0,
      1.0
    );

    // Write the resulting object out in pure JS format in case we want to paste it into
    // the ES console
    expect.addSnapshotSerializer({
      test: (val: any) => true,
      print: (val: any) => JSON.stringify(val, null, 3),
    });
    expect(test).toMatchSnapshot();
  });
});
