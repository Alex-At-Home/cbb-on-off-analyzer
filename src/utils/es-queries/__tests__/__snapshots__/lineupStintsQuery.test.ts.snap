// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lineupStintsQuery lineupStintsQuery 1`] = `
{
   "runtime_mappings": {
      "vs_high_major": {
         "type": "boolean",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value & 1) > 0)"
         }
      },
      "in_conf": {
         "type": "boolean",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value & 2) > 0)"
         }
      },
      "vs_rank": {
         "type": "long",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit((doc['common_lookup'].value >> 2) & 511)"
         }
      },
      "vs_3p": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) {\\n            def _3p = (doc['common_lookup'].value >> 11)  & 1023;\\n            if (_3p > 0) emit(0.1*_3p);\\n          }"
         }
      },
      "vs_adj_off": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit(0.1*((doc['common_lookup'].value >> 21) & 2047))"
         }
      },
      "vs_adj_def": {
         "type": "double",
         "script": {
            "source": "if (0 != doc['common_lookup'].size()) emit(0.1*((doc['common_lookup'].value >> 32) & 2047))"
         }
      },
      "common_lookup": {
         "type": "long",
         "script": {
            "source": "\\n            if (!params.kp_info.isEmpty()) {\\n              def kp_name = params.pbp_to_kp[doc[params.field_name].value];\\n              if (kp_name == null) {\\n                 kp_name = doc[params.field_name].value;\\n              } else {\\n                 kp_name = kp_name.pbp_kp_team;\\n              }\\n              def oppo = params.kp_info[kp_name];\\n              if (oppo != null) {\\n                def vs_high_major = oppo[\\"is_high_major\\"] ?: 1;\\n                if (vs_high_major instanceof String) {\\n                  vs_high_major = 0;\\n                }\\n                def oppo_conf = oppo[\\"conf\\"];\\n                def in_conf = params.conf.equals(oppo_conf);\\n                def hca = 0;\\n                if (doc[\\"location_type.keyword\\"].size() > 0) {\\n                  if (doc[\\"location_type.keyword\\"].value == \\"Neutral\\") {\\n                    hca = 25;\\n                  } else if (doc[\\"location_type.keyword\\"].value == \\"Away\\") {\\n                    hca = 50;\\n                  }\\n                }\\n                def margin_rank = Math.max(1, (oppo[\\"stats.adj_margin.rank\\"] ?: 0) - hca);\\n\\n                def _3p = oppo[\\"stats.off._3p_pct.value\\"] ?: 0.0;\\n                _3p = (_3p * 10).longValue() & 1023;\\n\\n                def adj_off = oppo[\\"stats.adj_off.value\\"] ?: 0;\\n                adj_off = (adj_off * 10).longValue() & 2047;\\n\\n                def adj_def = oppo[\\"stats.adj_def.value\\"] ?: 0;\\n                adj_def = (adj_def * 10).longValue() & 2047;\\n\\n                long returnVal = 0L | (vs_high_major << 0)\\n                  | ((in_conf ? 1 : 0) << 1)\\n                  | ((margin_rank.longValue() & 511) << 2)\\n                  | (_3p.longValue() << 11)\\n                  | (adj_off.longValue() << 21)\\n                  | (adj_def.longValue() << 32)\\n                  ;\\n                emit(returnVal);\\n              }\\n            }\\n          ",
            "lang": "painless",
            "params": {
               "pbp_to_kp": {
                  "name1": "name1b"
               },
               "field_name": "opponent.team.keyword",
               "kp_info": {
                  "team": {
                     "stats": 0
                  }
               },
               "conf": ""
            }
         }
      }
   },
   "_source": {
      "includes": [],
      "excludes": [
         "agent",
         "host",
         "log",
         "ecs"
      ]
   },
   "sort": [
      "start_min"
   ],
   "size": 200,
   "query": {
      "bool": {
         "filter": [],
         "minimum_should_match": 0,
         "should": [],
         "must_not": [],
         "must": [
            {
               "bool": {
                  "filter": [],
                  "must_not": [],
                  "should": [],
                  "minimum_should_match": 0,
                  "must": [
                     {
                        "term": {
                           "team.team.keyword": "TestTeam"
                        }
                     },
                     {
                        "query_string": {
                           "query": "vs_rank:[10 TO 100]"
                        }
                     }
                  ]
               }
            },
            {
               "query_string": {
                  "query": "players.id:(base)"
               }
            }
         ]
      }
   }
}
`;
